### ISSUE TITLE
Sampling fails in discrete models

### ISSUE THREAD

**Issue by davidoj** [pyro-ppl/numpyro#1518]
I'm using numpyro version 0.10.1 and funsor version 0.4.3, and I'm unable to sample a model with discrete variables.

Minimal working example:

```python
import numpyro
from numpyro.infer import MCMC, NUTS
import numpyro.distributions as dist
from jax import random


def discrete_model():
    X = numpyro.sample("X", dist.BernoulliProbs(0.5), infer={"enumerate": "parallel"})

mcmc = MCMC(
        NUTS(discrete_model),
        num_warmup=1000,
        num_samples=2000
    )
mcmc.run(random.PRNGKey(0))
mcmc.print_summary()
```

This results in the following error:
```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[5], line 15
      8     X = numpyro.sample("X", dist.BernoulliProbs(0.5), infer={"enumerate": "parallel"})
     10 mcmc = MCMC(
     11         NUTS(discrete_model),
     12         num_warmup=1000,
     13         num_samples=2000
     14     )
---> 15 mcmc.run(random.PRNGKey(0))
     16 mcmc.print_summary()
```
[...]
```python
File /opt/homebrew/lib/python3.10/site-packages/funsor/tensor.py:139, in Tensor.__init__(self, data, inputs, dtype)
    138 def __init__(self, data, inputs=None, dtype="real"):
--> 139     assert ops.is_numeric_array(data)
    140     assert isinstance(inputs, tuple)
    141     if not get_tracing_state():

AssertionError: 
```

I also tried running the tests in `tests/contrib/test_funsor.py` and `tests/contrib/test_infer_discrete.py`. They all fail with the same assertion error, with an additional message

```python
self = <[AttributeError("'Tensor' object has no attribute 'data'") raised in repr()] Tensor object at 0x168349bd0>
data = Array([0, 1], dtype=int32), inputs = (('z1', Bint[2]),), dtype = 2

    def __init__(self, data, inputs=None, dtype="real"):
>       assert ops.is_numeric_array(data)
E       AssertionError
```

**Comment by ordabayevy** [comment]
I believe this is due to the new version of JAX 0.4 and has been fixed in PR https://github.com/pyro-ppl/funsor/pull/601 in the master branch of funsor but not in the release version yet.

**Comment by davidoj** [comment]
Thanks, downgrading jax and jaxlib to 0.3.25 worked.

**Comment by fehiepsi** [comment]
We just released funsor 0.4.4, which should resolve the issue. Please consider updating jax and funsor to the latest release. Thanks for reporting!
